File: packages/gg-sandbox-ruby/src/gg/context.rb
# frozen_string_literal: true

module GG
  # Manage the agent's own context window.
  #
  # These are the only calls whose effect is on the conversation rather than on the workspace. They
  # are worth making from a program precisely because a program can decide *when* to: read a set of
  # files, extract what matters, then evict the views, all in one turn.
  module Context
    extend Surface::Operations

    # The membrane's reclaim record, as the model-facing one.
    #
    # @param report [Object] the wire record
    # @return [GG::Context::ReclaimReport] what the reclaim freed
    # @api private
    def self.report(report)
      ReclaimReport.new(
        items: Wire.field(report, "items"),
        reclaimed_tokens: Wire.field(report, "reclaimedTokens"),
        paths: Wire.field(report, "paths"),
        detail: Wire.field(report, "detail")
      )
    end
    private_class_method :report

    # Drop the contents of files that were read out of the context window, freeing their tokens.
    #
    # The files on disk are untouched: this forgets what was read, not what exists.
    #
    # @param path [String, nil] The file whose views to drop. Leave it out to drop every file view
    #   held.
    # @return [GG::Context::ReclaimReport] what the eviction freed
    # @raise [GG::Core::ToolError] `:not_found` when the named path has no view open.
    def self.evict_file_view(path = nil)
      report(Wire.call("evict_file_view", "context", "evictFileView", [Wire.js(path)]))
    end
    operation :evict_file_view, "context.evict_file_view", tool: "evict_file_view"

    # Move whole turns out of the context window and report what that reclaimed.
    #
    # Every result carries a header with its turn number and roughly what holding it costs, which is
    # what names the turns worth dropping. A span is a Ruby `Range` and both ends are included, so
    # `GG::Context.archive_thread(4..19)` archives turns 4 through 19; an exclusive range means the
    # same thing and is read the same way. The agent's own messages in an archived turn are dropped;
    # the results are kept and stay searchable with `GG::Context.search_archive`.
    #
    # @param ranges [Array<GG::Context::TurnRange>] The inclusive spans of turn numbers to move out
    #   of the window, splatted: `GG::Context.archive_thread(4..19, 30..35)`.
    # @return [GG::Context::ReclaimReport] what the archive freed
    # @raise [GG::Core::ToolError] `:invalid_argument` for anything that is not a range of turn
    #   numbers, for an empty list, and for too many spans at once.
    def self.archive_thread(*ranges)
      spans = ranges.flatten.map do |span|
        first, last = Check.span("archive_thread", span)
        Wire.record("start" => first, "end" => last)
      end
      report(Wire.call("archive_thread", "context", "archiveThread", [spans]))
    end
    operation :archive_thread, "context.archive_thread", tool: "archive_thread"

    # Search archived history for a case-insensitive substring, most recent first, up to 8 hits.
    #
    # `archive_empty?` is worth checking before `hits`: it distinguishes "nothing has been archived
    # yet" from "the search ran and matched nothing", so a program does not archive again believing
    # the first archive failed.
    #
    # @param query [String] The substring to look for. Matching is case-insensitive.
    # @return [GG::Context::ArchiveSearch] whether anything is archived at all, and what matched
    # @raise [GG::Core::ToolError] `:invalid_argument` for a blank query.
    def self.search_archive(query)
      found = Wire.call("search_archive", "context", "searchArchive", [query])
      ArchiveSearch.new(
        archive_empty: Wire.field(found, "archiveEmpty"),
        hits: Wire.field(found, "hits").map do |hit|
          ArchiveHit.new(
            seq: Wire.field(hit, "seq"),
            role: Wire.symbol(`#{hit}.role`),
            text: Wire.field(hit, "text")
          )
        end
      )
    end
    operation :search_archive, "context.search_archive", tool: "search_archive"

    # Compact the context window: the detailed thread is dropped and restarted from `summary`.
    #
    # A fresh read of each path in `files` is added to the restarted window. Skills, memories and
    # the task list are kept as they are. gg asks for this call when the window is full, and refuses
    # every other call until it arrives.
    #
    # It does not stop the program: it registers the request and returns, and the rewrite happens
    # once the program has ended. Everything not in the summary and not in `files` is gone, so the
    # summary is written for the agent that comes after and `files` names what it will need in hand.
    #
    # @param summary [String] What the restarted window opens with. Everything not in it and not
    #   re-read from `files` is gone.
    # @param files [Array<String>] The paths to read afresh into the restarted window. Defaults to
    #   none.
    # @return [nil] nothing; the request is registered and the program runs on
    # @raise [GG::Core::ToolError] `:invalid_argument` for a blank summary.
    def self.compact(summary, files: [])
      Wire.call("compact", "context", "compact",
                [summary, Check.strings("compact", "files", files)])
      nil
    end
    operation :compact, "context.compact", tool: "compact"

    # What a context reclaim actually freed from the live context window.
    class ReclaimReport
      include Value

      # @return [Integer] Context items dropped from the live window.
      attr_reader :items

      # @return [Integer] Approximately how many tokens that freed.
      attr_reader :reclaimed_tokens

      # @return [Array<String>] The workspace paths whose views were evicted. Empty for an archive.
      attr_reader :paths

      # @return [String] The prose summary of what was reclaimed.
      attr_reader :detail

      # @api private
      def initialize(items:, reclaimed_tokens:, paths:, detail:)
        @items = items
        @reclaimed_tokens = reclaimed_tokens
        @paths = paths
        @detail = detail
        freeze
      end
    end

    # An inclusive span of turn numbers, which is Ruby's own `Range`.
    #
    # The numbers are the ones on the header of every result, so `4..19` means exactly the turns
    # numbered 4 through 19, both ends included. An exclusive range (`4...20`) means the same thing
    # and is read the same way.
    TurnRange = ::Range

    # Who said an archived message.
    #
    # Every arm is a Symbol, so a comparison may name the constant or write the literal:
    # `GG::Context::MessageRole::USER` and `:user` are the same value.
    module MessageRole
      # The system prompt.
      SYSTEM = :system

      # A turn's input to the agent — a result, a view, or an operator's instruction.
      USER = :user

      # Something the agent said.
      ASSISTANT = :assistant

      # A tool result, on a session that made tool calls rather than writing programs.
      TOOL = :tool
    end

    # One archived message that matched a search.
    class ArchiveHit
      include Value

      # @return [Integer] The archived message's sequence number.
      attr_reader :seq

      # @return [GG::Context::MessageRole] Who said it.
      attr_reader :role

      # @return [String] The message text.
      attr_reader :text

      # @api private
      def initialize(seq:, role:, text:)
        @seq = seq
        @role = role
        @text = text
        freeze
      end
    end

    # What `GG::Context.search_archive` found.
    class ArchiveSearch
      include Value

      # @return [Array<GG::Context::ArchiveHit>] The matches, most recent first, at most 8.
      attr_reader :hits

      # @api private
      def initialize(archive_empty:, hits:)
        @archive_empty = archive_empty
        @hits = hits
        freeze
      end

      # Nothing has been archived yet, so there was nothing to search.
      #
      # Deliberately distinct from a search that ran and matched nothing, so a program does not
      # archive again believing the first archive failed.
      #
      # @return [Boolean]
      def archive_empty?
        @archive_empty
      end
    end
  end
end

File: packages/gg-sandbox-python/src/gg/context.py
"""Manage the agent's own context window.

These are the only calls whose effect is on the conversation rather than on the workspace. They are
worth making from a program precisely because a program can decide *when* to: read a set of files,
extract what matters, then evict the views, all in one turn.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from wit_world.imports import context as wire

from ._registry import missing, operation
from .core import ToolError, ToolErrorCode, _call, _strings, _uint

__all__ = [
    "ArchiveHit",
    "ArchiveSearch",
    "MessageRole",
    "ReclaimReport",
    "TurnRange",
    "archive_thread",
    "compact",
    "evict_file_view",
    "search_archive",
]


@dataclass(frozen=True)
class ReclaimReport:
    """What a context reclaim actually freed from the live context window."""

    items: int
    """Context items dropped from the live window."""

    reclaimed_tokens: int
    """Approximately how many tokens that freed."""

    paths: list[str]
    """The workspace paths whose views were evicted. Empty for an archive."""

    detail: str
    """The prose summary of what was reclaimed."""


@dataclass(frozen=True)
class TurnRange:
    """An inclusive span of turn numbers, the unit `archive_thread` moves out of the window.

    The numbers are the ones on the header of every result, so `TurnRange(4, 19)` means exactly the
    turns numbered 4 through 19 — both ends included.
    """

    start: int
    """The first turn in the span."""

    end: int
    """The last turn in the span, inclusive."""


class MessageRole(Enum):
    """Who said an archived message."""

    SYSTEM = "system"
    """The system prompt."""

    USER = "user"
    """A turn's input to the agent — a result, a view, or an operator's instruction."""

    ASSISTANT = "assistant"
    """Something the agent said."""

    TOOL = "tool"
    """A tool result, on a session that made tool calls rather than writing programs."""


@dataclass(frozen=True)
class ArchiveHit:
    """One archived message that matched a search."""

    seq: int
    """The archived message's sequence number."""

    role: MessageRole
    """Who said it."""

    text: str
    """The message text."""


@dataclass(frozen=True)
class ArchiveSearch:
    """What `search_archive` found."""

    archive_empty: bool
    """Nothing has been archived yet, so there was nothing to search.

    Deliberately distinct from a search that ran and matched nothing, so a program does not archive
    again believing the first archive failed.
    """

    hits: list[ArchiveHit]
    """The matches, most recent first, at most 8."""


def _report(report: wire.ReclaimReport) -> ReclaimReport:
    """The membrane's reclaim record, as the model-facing one."""
    return ReclaimReport(
        items=report.items,
        reclaimed_tokens=report.reclaimed_tokens,
        paths=[*report.paths],
        detail=report.detail,
    )


def _span(span: TurnRange, edge: str) -> int:
    """One end of a turn span, checked before it is lowered.

    A `TurnRange` is a plain dataclass, so nothing stops a program constructing one out of a string
    or a negative number; the membrane would lower either into a `u32` without complaint, and archive
    a range nobody asked for.
    """
    if not isinstance(span, TurnRange):
        raise ToolError(
            "archive_thread",
            ToolErrorCode.INVALID_ARGUMENT,
            f"every entry of `ranges` must be a TurnRange, got {span!r}",
        )
    value = _uint("archive_thread", f"ranges[].{edge}", getattr(span, edge))
    if value is None:
        raise ToolError(
            "archive_thread",
            ToolErrorCode.INVALID_ARGUMENT,
            f"every entry of `ranges` needs both `start` and `end`; `{edge}` is None",
        )
    return value


@operation("context.evict_file_view")
def evict_file_view(path: str | None = None) -> ReclaimReport:
    """Drop the contents of files that were read out of the context window.

    The files on disk are untouched: this forgets what was read, not what exists.

    Args:
        path: The file whose views to drop. The default drops every file view held.

    Returns:
        How many context items went and roughly how many tokens that reclaimed, with `paths` naming
            the files whose views were dropped.

    Raises:
        ToolError: `invalid-argument` for a path that is given but empty; the default is how every
            file view is dropped.
    """
    return _report(_call(wire.evict_file_view, path))


@operation("context.archive_thread")
def archive_thread(ranges: list[TurnRange]) -> ReclaimReport:
    """Move whole turns out of the context window.

    Every result carries a header with its turn number and roughly what holding it costs, which is
    what names the turns worth dropping: `gg.context.archive_thread([gg.context.TurnRange(4, 19)])`
    archives turns 4 through 19, both ends included. The agent's own messages in an archived turn are
    dropped; the results are kept and stay searchable with `search_archive`.

    Args:
        ranges: The inclusive spans of turn numbers to move out of the window. They may overlap.

    Returns:
        How many context items went and roughly how many tokens that reclaimed. `paths` is empty,
            since an archive drops turns rather than files.

    Raises:
        ToolError: `invalid-argument` for an empty list, too many spans at once, or a span that ends
            before it starts.
    """
    spans = [wire.TurnRange(start=_span(span, "start"), end=_span(span, "end")) for span in ranges]
    return _report(_call(wire.archive_thread, spans))


@operation("context.search_archive")
def search_archive(query: str) -> ArchiveSearch:
    """Search archived history for a case-insensitive substring, most recent first, up to 8 hits.

    `archive_empty` is worth checking before `hits`: it distinguishes "nothing has been archived yet"
    from "the search ran and matched nothing", so a program does not archive again believing the
    first archive failed.

    Args:
        query: The substring to look for. Matching is case-insensitive.

    Returns:
        The matches, most recent first and at most 8, beside the `archive_empty` flag.

    Raises:
        ToolError: `invalid-argument` for an empty query.
    """
    found = _call(wire.search_archive, query)
    return ArchiveSearch(
        archive_empty=found.archive_empty,
        hits=[
            ArchiveHit(seq=hit.seq, role=MessageRole[hit.role.name], text=hit.text)
            for hit in found.hits
        ],
    )


@operation("context.compact")
def compact(summary: str, files: list[str] | None = None) -> None:
    """Compact the context window: the detailed thread is dropped and restarted from `summary`.

    A fresh read of each path in `files` is added to the restarted window. Skills, memories and the
    task list are kept as they are. gg asks for this call when the window is full, and refuses every
    other call until it arrives.

    It does not stop the program: it registers the request and returns, and the rewrite happens once
    the program has ended. Everything not in the summary and not in `files` is gone, so the summary
    is written for the agent that comes after and `files` names what it will need in hand.

    Args:
        summary: What the restarted window opens with. Everything not in it and not re-read from
            `files` is gone.
        files: The paths to read afresh into the restarted window. The default reads nothing back.

    Raises:
        ToolError: `invalid-argument` for a blank summary. This is the one call gg does not refuse
            while a compaction is in flight, since nothing else can clear the window.
    """
    _call(wire.compact, summary, _strings("compact", "files", files))


__getattr__ = missing(__name__, __all__)
"""What this module answers for a name it does not declare — see `gg._registry.missing`."""
